tests/components/media_player/test_cast.py

```diff
@@ -18,83 +21,221 @@ def cast_mock():
-            'host': 'some_host',
-            'port': cast.DEFAULT_PORT,
-        })
-        assert not mock_device.called
-
-    @patch('homeassistant.components.media_player.cast.CastDevice')
-    @patch('pychromecast.get_chromecasts')
-    @patch('pychromecast.Chromecast')
-    def test_fallback_cast(self, mock_chromecast, mock_get_chromecasts,
-                           mock_device):
-        """Test falling back to creating Chromecast when not discovered."""
-        mock_get_chromecasts.return_value = [
-            FakeChromeCast('some_host', cast.DEFAULT_PORT)
-        ]
-
-        # Test chromecasts as if they were hardcoded in configuration.yaml
-        cast.setup_platform(self.hass, {
-            'host': 'some_other_host'
-        }, lambda _: _)
-
-        assert mock_chromecast.called
-        assert mock_device.called
-
-    @patch('homeassistant.components.media_player.cast.CastDevice')
-    @patch('pychromecast.get_chromecasts')
-    @patch('pychromecast.Chromecast')
-    def test_fallback_cast_group(self, mock_chromecast, mock_get_chromecasts,
-                                 mock_device):
-        """Test not creating Cast Group when not discovered."""
-        mock_get_chromecasts.return_value = [
-            FakeChromeCast('some_host', cast.DEFAULT_PORT)
-        ]
-
-        # Test chromecasts as if they were automatically discovered
-        cast.setup_platform(self.hass, {}, lambda _: _, {
-            'host': 'some_other_host',
-            'port': 43546,
-        })
-        assert not mock_chromecast.called
-        assert not mock_device.called
+# pylint: disable=invalid-name
+FakeUUID = UUID('57355bce-9364-4aa6-ac1e-eb849dccf9e2')
+
+
+def get_fake_chromecast(host='192.168.178.42', port=8009,
+                        uuid: Optional[UUID] = FakeUUID):
+    """Generate a Fake Chromecast object with the specified arguments."""
+    return MagicMock(host=host, port=port, uuid=uuid)
+
+
+@asyncio.coroutine
+def async_setup_cast(hass, config=None, discovery_info=None):
+    """Helper to setup the cast platform."""
+    if config is None:
+        config = {}
+    add_devices = Mock()
+
+    yield from cast.async_setup_platform(hass, config, add_devices,
+                                         discovery_info=discovery_info)
+    yield from hass.async_block_till_done()
+
+    return add_devices
+
+
+@asyncio.coroutine
+def async_setup_cast_internal_discovery(hass, config=None,
+                                        discovery_info=None,
+                                        no_from_host_patch=False):
+    """Setup the cast platform and the discovery."""
+    listener = MagicMock(services={})
+
+    with patch('pychromecast.start_discovery',
+               return_value=(listener, None)) as start_discovery:
+        add_devices = yield from async_setup_cast(hass, config, discovery_info)
+        yield from hass.async_block_till_done()
+        yield from hass.async_block_till_done()
```

Comment:
The reason you have to do this twice is because `@callback` are scheduled on the event loop without being wrapped in a task and thus cannot be tracked by `async_block_till_done`

Reply:
By doing another yield, callback is executed. If it then schedules a coroutine, that will get picked up again by block till done.